### PR TITLE
feat: adding the audio modality 🎤 

### PR DESCRIPTION
**Pull request by CarolinePascal**
## Adding audio modality in recordings and datasets

This PR introduces the audio modality in LeRobot. It allows users to record audio along with video using laptop-embedded or USB microphones, and to store the gathered data in a LeRobotDataset. It includes :
- A new `Microphone` class, inspired by the `Camera` class, which leverages the [sounddevice](https://python-sounddevice.readthedocs.io/) and [soundfile](https://python-soundfile.readthedocs.io) python packages to manage microphones and record multi-channel audio data. This class allows to record sound in two ways : (1) as a continuous signal directly written in a file over an extended period of time (e.g. an episode) or (2) as smaller 'audio chunks', which provide a snapshot of the past few second (e.g. for policy inputs). Both recording methods are running in a separate thread which should not impact performance.
- An integration of the `Microphone` class in the `Robot` class to easily define microphones and link them to cameras. By doing so, audio will be saved along with the recorded video in a single MP4 file;
- An integration of the audio modality  in LeRobotDataset, as a specific feature stored in a separated file (MP4 for video+audio, M4A for audio). Similarly to video files, audio files are encoded using `ffmpeg`. For now, only [torchaudio](https://pytorch.org/audio/stable/index.html) decoding is supported ([torchcodec](https://pytorch.org/torchcodec/main/generated_examples/audio_decoding.html#sphx-glr-generated-examples-audio-decoding-py) decoding is a work in progress). Once decoded, 'audio chunks' of fixed length may be queried based on a user given timestamp;
- A new set of dedicated tests.

## How it was tested

The feature was tested on macOS Sequoia with a SO100 setup with two cameras (laptop and phone), and two microphones (laptop and headset). One of the microphones was linked to a camera, and the other one treated as a standalone sensor. See the SO100 entry in the [config.py](https://github.com/CarolinePascal/lerobot/blob/user/CarolinePascal/2025_03_24_audio_dataset/lerobot/common/robot_devices/robots/configs.py) for more details.

## How to checkout & try? 

_Assuming a SO100 setup_

1) Install audio dependencies 

```
pip install -e ".[audio]"
```

2) Query microphones indices and update the robot `config.py` file

```
python -m sounddevice
```
_The microphones have at least one input channel_

You may get the microphone sample rate as follows:

```
python -c "import sounddevice as sd; print(sd.query_devices(<microphone_index>)['default_samplerate'])"
```

3) Run a dummy dataset recording

```
python lerobot/scripts/control_robot.py \
  --robot.type=so100 \
  --control.type=record \
  --control.fps=30 \
  --control.single_task="Hear some sound." \
  --control.repo_id=so100/sound_test \
  --control.tags='["so100","sound"]' \
  --control.warmup_time_s=2 \
  --control.episode_time_s=5 \
  --control.reset_time_s=1 \
  --control.num_episodes=1 \
  --control.push_to_hub=false
```

You should be able to check the recorded data in your local dataset folder.

### CODE REVIEW

**Comment by CarolinePascal** [comment]
⚠️ Running tests highlighted that `tochaudio` (used for audio decoding, as `torchcodec` does not officially supports it yet) if not compatible with `ffmpeg > 7`, which is now the recommended version.

There is a working fix on *macOS* which relies on installing `ffmpeg` with :

```
conda install ffmpeg=6.1.2 -c conda-forge
```

-> No conflict with the latest versions of `av` or `torchvision` (which must encapsulate their own `ffmpeg` version);
-> Compatible with `torchaudio` (<7) and `torchcodec` (installed with conda);
-> Already lists `libsvtav1` among its encoders.

**Review by Cadene — COMMENTED** [review]
Super nice work ;) Almost mergeable! After our discussion I wrote this little TODO. I hope it helps!

TODO:
- test record microphone at training time versus at inference 
- find a small audio model to add to ACT (no need to merge this part for now) + train a policy and evaluate it as soon as possible ;D
- simplify code with focus on m4a instead of mp4
- (when possible) simplify code with focus on torchcodec instead of torchaudio

File: lerobot/common/datasets/video_utils.py [hunk 1]
```diff
@@ -40,6 +42,74 @@ def get_safe_default_codec():
         return "pyav"
 
 
+def decode_audio(
```
[the remark this answers is not in the document]
Comment (CarolinePascal) [reply lerobot/common/datasets/video_utils.py]: Done in [ca716ed](https://github.com/huggingface/lerobot/pull/967/commits/ca716ed1960aecc5da835264e0313db477e68ca1) ! 
More generally, this commit completely splits video and audio handling (separate encoding and storage, no links between `Camera` and `Microphone` classes).

File: lerobot/common/datasets/lerobot_dataset.py [hunk 2]
```diff
@@ -904,6 +1093,13 @@ def save_episode(self, episode_data: dict | None = None) -> None:
         if img_dir.is_dir():
             shutil.rmtree(self.root / "images")
 
+        # delete raw audio files
+        raw_audio_files = list(self.root.rglob("*.wav"))
+        for raw_audio_file in raw_audio_files:
+            raw_audio_file.unlink()
+            if len(list(raw_audio_file.parent.iterdir())) == 0:
+                raw_audio_file.parent.rmdir()
```
[the remark this answers is not in the document]
Comment (CarolinePascal) [reply lerobot/common/datasets/lerobot_dataset.py:1082]: This would also delete the encoded m4a files et this point :/

[the remark this answers is not in the document]
Comment (CarolinePascal) [reply lerobot/common/datasets/lerobot_dataset.py]: Done in [6120276](https://github.com/huggingface/lerobot/pull/967/commits/61202761f5fd95937e9cd0ecff73c56aceeeed5d) !

File: lerobot/common/robot_devices/control_utils.py [hunk 3]
```diff
@@ -107,11 +118,15 @@ def predict_action(observation, policy, device, use_amp):
...
-        # Convert to pytorch format: channel first and float32 in [0,1] with batch dimension
         for name in observation:
+            # Convert to pytorch format: channel first and float32 in [0,1] with batch dimension
             if "image" in name:
                 observation[name] = observation[name].type(torch.float32) / 255
                 observation[name] = observation[name].permute(2, 0, 1).contiguous()
+            # Convert to pytorch format: channel first and float32 in [-1,1] with batch dimension
+            if "audio" in name:
+                observation[name] = observation[name].type(torch.float32)
+                observation[name] = observation[name].permute(1, 0).contiguous()
```
Comment (milong26) [inline lerobot/common/robot_devices/control_utils.py:129]: How to convert to pytorch format: channel first and float32 in [-1,1] with batch dimension?

We know images' range(0,255) ,so should we know audio's range?

Comment (CarolinePascal) [reply lerobot/common/robot_devices/control_utils.py:129]: This is actually the expected range of audio data when audio samples are loaded as floats. With the current implemenation, the `float32` conversion will actually not do much, as input audio samples are already in the correct format.
The actual normalisation needs to be handled in the policy according to the pre-trained model stats (see ACT implemenation in this branch).
